### PR TITLE
Distil gpt neo

### PR DESCRIPTION
**Pull request by preethamgali**
Distillation script for GPTNeo . Please check the working as I didn't have sufficient GPU to check the working.

### CODE REVIEW

**Review by StellaAthena — CHANGES_REQUESTED** [review]
This throws a weird error. The code doesn’t seem to like something you’re doing with the `os` package? You’re trying to manipulate file structures with a function that doesn’t exist.

Even without the local GPU to do the fine-tuning, you should be able to get it to run. It’s probably also worth trying it out on [the 125M model](https://huggingface.co/EleutherAI/gpt-neo-125M) if that fits locally.

I’m expecting to have spare compute to run this either this weekend or next week. Right now all the A100s are spoken for as we are rushing something for NeurIPS, but there’ll be free compute soon.

**Comment by kinoc** [comment]
Testing the branch on the local TITAN (about 3090 equiv)
To get it to run locally I had to make the following mods:

- In create_data.py Change "os.list_dir" to "os.listdir"

- assertions for the hidden_size and hidden_states.size being the same fail since 1.3B and 125M/350M have different hidden state sizes. So in run_distillation.py comment out line 94 assert: ```#    assert student.config.hidden_size == teacher.config.hidden_size```

- Used an old copy of StackOverflow AI (about 500K) as as a quick debugging training text file, and placed it in ```./train```
- Created ```train_data.pkl``` file using ```python3 create_data.py --train_files_path ./train```

To run the system on the TITAN used the following command line:
```python3 run_distillation.py --batch_size 2 --alpha_cos 0.0 --dump_path s1 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-125M```
or
```python3 run_distillation.py --batch_size 1 --n_epoch 45 --alpha_cos 0.0 --alpha_ce 5.0  --alpha_clm 2.5 --alpha_mse 1.0 --dump_path s350 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-350M```

--batch_size 2 to get past the OOM
--alpha_cos 0.0 to get past distiller.py line 440 : ```assert s_hidden_states.size() == t_hidden_states.size()```

Param: Namespace(adam_epsilon=1e-06, alpha_ce=5.0, alpha_clm=2.5, alpha_cos=0.0, alpha_mlm=0.0, alpha_mse=0.5, batch_size=1, checkpoint_interval=1500, data_file='train_data.pkl', dump_path='s1', force=True, fp16=False, fp16_opt_level='O1', freeze_pos_embs=False, global_rank=0, gradient_accumulation_steps=50, group_by_size=True, initializer_range=0.02, is_master=True, learning_rate=0.0005, local_rank=0, log_interval=500, max_grad_norm=5.0, mlm=None, multi_gpu=False, multi_node=False, n_epoch=16, n_gpu=1, n_gpu_per_node=1, n_nodes=1, node_id=0, restrict_ce_to_mask=False, seed=2020, student_config='EleutherAI/gpt-neo-350M', student_pretrained_weights=None, teacher_name='EleutherAI/gpt-neo-1.3B', temperature=1.5, warmup_prop=0.05, weight_decay=0.0, world_size=1)

Using teacher_name='EleutherAI/gpt-neo-1.3B'
batch_size=1 , gpt-neo-125M  (learning)  => 12644MiB / 24190MiB
batch_size=2, gpt-neo-125M  (not learning) =>18990MiB / 24190MiB
batch_size=3, gpt-neo-125M  (not learning) => 23928MiB / 24190MiB

batch_size=1, gpt-neo-350M  (learning) => 17292MiB / 24190MiB
batch_size=2, gpt-neo-350M  (---) => out-of-memory/ 24190MiB

Questions about the various alpha values and effect of setting alpha_cos to 0.0. Loss does make a steady decrease, and runs to completion (unless you of course you run out of disk space during snapshots ...). Need to check why loss does not decrease for for batch_size>1.

**Comment by kinoc** [comment]
Two minor additions that would be nice:
- Add to distiller.py after 187  ```self.tensorboard.add_text(tag="config/teacher", text_string=str(self.teacher.config), global_step=0)``` to see the teacher config in tensorboard and see the diff between models
- Add 'restore from last checkpoint if present' as an option

**Comment by StellaAthena** [comment]
> Testing the branch on the local TITAN (about 3090 equiv)
> To get it to run locally I had to make the following mods:
> 
>     * In create_data.py Change "os.list_dir" to "os.listdir"
> 
>     * assertions for the hidden_size and hidden_states.size being the same fail since 1.3B and 125M/350M have different hidden state sizes. So in run_distillation.py comment out line 94 assert: `#    assert student.config.hidden_size == teacher.config.hidden_size`
> 
>     * Used an old copy of StackOverflow AI (about 500K) as as a quick debugging training text file, and placed it in `./train`
> 
>     * Created `train_data.pkl` file using `python3 create_data.py --train_files_path ./train`
> 
> 
> To run the system on the TITAN used the following command line:
> `python3 run_distillation.py --batch_size 2 --alpha_cos 0.0 --dump_path s1 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-125M`
> or
> `python3 run_distillation.py --batch_size 1 --n_epoch 45 --alpha_cos 0.0 --alpha_ce 5.0 --alpha_clm 2.5 --alpha_mse 1.0 --dump_path s350 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-350M`
> 
> --batch_size 2 to get past the OOM
> --alpha_cos 0.0 to get past distiller.py line 440 : `assert s_hidden_states.size() == t_hidden_states.size()`
> 
> Param: Namespace(adam_epsilon=1e-06, alpha_ce=5.0, alpha_clm=2.5, alpha_cos=0.0, alpha_mlm=0.0, alpha_mse=0.5, batch_size=1, checkpoint_interval=1500, data_file='train_data.pkl', dump_path='s1', force=True, fp16=False, fp16_opt_level='O1', freeze_pos_embs=False, global_rank=0, gradient_accumulation_steps=50, group_by_size=True, initializer_range=0.02, is_master=True, learning_rate=0.0005, local_rank=0, log_interval=500, max_grad_norm=5.0, mlm=None, multi_gpu=False, multi_node=False, n_epoch=16, n_gpu=1, n_gpu_per_node=1, n_nodes=1, node_id=0, restrict_ce_to_mask=False, seed=2020, student_config='EleutherAI/gpt-neo-350M', student_pretrained_weights=None, teacher_name='EleutherAI/gpt-neo-1.3B', temperature=1.5, warmup_prop=0.05, weight_decay=0.0, world_size=1)
> 
> Using teacher_name='EleutherAI/gpt-neo-1.3B'
> batch_size=1 , gpt-neo-125M (learning) => 12644MiB / 24190MiB
> batch_size=2, gpt-neo-125M (not learning) =>18990MiB / 24190MiB
> batch_size=3, gpt-neo-125M (not learning) => 23928MiB / 24190MiB
> 
> batch_size=1, gpt-neo-350M (learning) => 17292MiB / 24190MiB
> batch_size=2, gpt-neo-350M (---) => out-of-memory/ 24190MiB
> 
> Questions about the various alpha values and effect of setting alpha_cos to 0.0. Loss does make a steady decrease, and runs to completion (unless you of course you run out of disk space during snapshots ...). Need to check why loss does not decrease for for batch_size>1.

Would you mind pushing the code you got working locally?

**Comment by StellaAthena** [comment]
@preethamgali any update on this?

**Comment by preethamgali** [comment]
> @preethamgali any update on this?

Please give me time to review. I will check this out soon.

**Comment by preethamgali** [comment]
> Two minor additions that would be nice:
> 
> * Add to distiller.py after 187  `self.tensorboard.add_text(tag="config/teacher", text_string=str(self.teacher.config), global_step=0)` to see the teacher config in tensorboard and see the diff between models
> * Add 'restore from last checkpoint if present' as an option
> 
> gpt-neo-1.3B hidden_size of 2048 != gpt-neo-350M hidden_size of 1024
> 
> Teacher config (gpt-neo-1.3B):
> GPTNeoConfig { "_name_or_path": "EleutherAI/gpt-neo-1.3B", "activation_function": "gelu_new", "architectures": [ "GPTNeoForCausalLM" ], "attention_dropout": 0, "attention_layers": [ "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local" ], "attention_types": [ [ [ "global", "local" ], 12 ] ], "bos_token_id": 50256, "embed_dropout": 0, "eos_token_id": 50256, "gradient_checkpointing": false, **"hidden_size": 2048**, "initializer_range": 0.02, "intermediate_size": null, "layer_norm_epsilon": 1e-05, "max_position_embeddings": 2048, "model_type": "gpt_neo", "num_heads": 16, "num_layers": 24, "output_hidden_states": true, "resid_dropout": 0, "summary_activation": null, "summary_first_dropout": 0.1, "summary_proj_to_labels": true, "summary_type": "cls_index", "summary_use_proj": true, "task_specific_params": { "text-generation": { "do_sample": true, "max_length": 50, "temperature": 0.9 } }, "tokenizer_class": "GPT2Tokenizer", "transformers_version": "4.5.0.dev0", "use_cache": true, "vocab_size": 50257, "window_size": 256 }
> 
> Student config (gpt-neo-350M):
> GPTNeoConfig { "activation_function": "gelu_new", "architectures": [ "GPTNeoForCausalLM" ], "attention_dropout": 0, "attention_layers": [ "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local", "global", "local" ], "attention_types": [ [ [ "global", "local" ], 12 ] ], "bos_token_id": 50256, "embed_dropout": 0, "eos_token_id": 50256, "gradient_checkpointing": false, **"hidden_size": 1024**, "initializer_range": 0.02, "intermediate_size": null, "layer_norm_epsilon": 1e-05, "max_position_embeddings": 2048, "model_type": "gpt_neo", "num_heads": 16, "num_layers": 24, "output_hidden_states": true, "resid_dropout": 0, "summary_activation": null, "summary_first_dropout": 0.1, "summary_proj_to_labels": true, "summary_type": "cls_index", "summary_use_proj": true, "transformers_version": "4.5.0.dev0", "use_cache": true, "vocab_size": 50257, "window_size": 256 }

✅ self.tensorboard.add_text(tag="config/teacher", text_string=str(self.teacher.config) 
Please use --student_pretrained_weights to set the path containing the trained model or checkpoint for 'restore from the last checkpoint if present'.

**Comment by preethamgali** [comment]
> Testing the branch on the local TITAN (about 3090 equiv)
> To get it to run locally I had to make the following mods:
> 
> * In create_data.py Change "os.list_dir" to "os.listdir"
> * assertions for the hidden_size and hidden_states.size being the same fail since 1.3B and 125M/350M have different hidden state sizes. So in run_distillation.py comment out line 94 assert: `#    assert student.config.hidden_size == teacher.config.hidden_size`
> * Used an old copy of StackOverflow AI (about 500K) as as a quick debugging training text file, and placed it in `./train`
> * Created `train_data.pkl` file using `python3 create_data.py --train_files_path ./train`
> 
> To run the system on the TITAN used the following command line:
> `python3 run_distillation.py --batch_size 2 --alpha_cos 0.0 --dump_path s1 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-125M`
> or
> `python3 run_distillation.py --batch_size 1 --n_epoch 45 --alpha_cos 0.0 --alpha_ce 5.0 --alpha_clm 2.5 --alpha_mse 1.0 --dump_path s350 --data_file train_data.pkl --student_config EleutherAI/gpt-neo-350M`
> 
> --batch_size 2 to get past the OOM
> --alpha_cos 0.0 to get past distiller.py line 440 : `assert s_hidden_states.size() == t_hidden_states.size()`
> 
> Param: Namespace(adam_epsilon=1e-06, alpha_ce=5.0, alpha_clm=2.5, alpha_cos=0.0, alpha_mlm=0.0, alpha_mse=0.5, batch_size=1, checkpoint_interval=1500, data_file='train_data.pkl', dump_path='s1', force=True, fp16=False, fp16_opt_level='O1', freeze_pos_embs=False, global_rank=0, gradient_accumulation_steps=50, group_by_size=True, initializer_range=0.02, is_master=True, learning_rate=0.0005, local_rank=0, log_interval=500, max_grad_norm=5.0, mlm=None, multi_gpu=False, multi_node=False, n_epoch=16, n_gpu=1, n_gpu_per_node=1, n_nodes=1, node_id=0, restrict_ce_to_mask=False, seed=2020, student_config='EleutherAI/gpt-neo-350M', student_pretrained_weights=None, teacher_name='EleutherAI/gpt-neo-1.3B', temperature=1.5, warmup_prop=0.05, weight_decay=0.0, world_size=1)
> 
> Using teacher_name='EleutherAI/gpt-neo-1.3B'
> batch_size=1 , gpt-neo-125M (learning) => 12644MiB / 24190MiB
> batch_size=2, gpt-neo-125M (not learning) =>18990MiB / 24190MiB
> batch_size=3, gpt-neo-125M (not learning) => 23928MiB / 24190MiB
> 
> batch_size=1, gpt-neo-350M (learning) => 17292MiB / 24190MiB
> batch_size=2, gpt-neo-350M (---) => out-of-memory/ 24190MiB
> 
> Questions about the various alpha values and effect of setting alpha_cos to 0.0. Loss does make a steady decrease, and runs to completion (unless you of course you run out of disk space during snapshots ...). Need to check why loss does not decrease for for batch_size>1.

For the alpha values:
alpha_ce : Linear weight for KLDivLoss between student and teacher token classification outputs.
alpha_clm : Linear weight for CrossEntropyLoss between stduent token classification outputs and the labels.
alpha_mse : Linear weight for MSELoss between student and teacher token classification logits.
alpha_cos : Linear weight for CosineEmbeddingLoss between student and teacher last layers hidden_states.(works only when the model hidden_size of student and teacher are same)

For traning:
I don't have enough GPU to test the model. I am not able to load both teacher and student on the GPU i have.

**Comment by StellaAthena** [comment]
Great! I'll test it out later today @preethamgali

**Comment by StellaAthena** [comment]
Okay, so the code appears to run but the preprocessing of the data is extremely slow. I tried preprocessing a chunk of the Pile and it told me it would take 9 hours to preprocess the 45 GB file. 5 GB/hour is not an acceptable speed, so we need to work on that.
